lightning/api_grpc.go

```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
+	}
+
+	return resp.Txid, nil
+}
+
+// PayInvoice API.
+func (l *LndGrpcLightningAPI) PayInvoice(ctx context.Context, paymentRequest string, sats int64, outgoingChanIds []uint64) (*PaymentResp, error) {
+	req := &routerrpc.SendPaymentRequest{}
+	req.PaymentRequest = paymentRequest
+	if sats > 0 {
+		req.Amt = sats
+	}
+
+	if outgoingChanIds != nil {
+		req.OutgoingChanIds = make([]uint64, 0)
+		req.OutgoingChanIds = append(req.OutgoingChanIds, outgoingChanIds...)
+	}
+
+	// TODO: this is mandatory field but timeout could be configurable
+	req.TimeoutSeconds = 10
+
+	resp, err := l.RouterClient.SendPaymentV2(ctx, req)
+
+	if err != nil {
+		if strings.Contains(err.Error(), "invoice is already paid") {
+			// TODO: we don't know the hash here, else we could return l.GetPaymentStatus()
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout")
+		default:
+			// Do nothing
+		}
+
+		event, err := resp.Recv()
+
+		if err != nil {
+			return nil, err
+		}
+
+		switch event.Status {
+		case lnrpc.Payment_SUCCEEDED:
+			return &PaymentResp{
+				Preimage: event.PaymentPreimage,
+				Hash:     event.PaymentHash,
+				Status:   Success,
+			}, nil
+
+		case lnrpc.Payment_IN_FLIGHT:
+			var htlcSum int64
+
+			for _, htlc := range event.Htlcs {
+				htlcSum += htlc.Route.TotalAmtMsat - htlc.Route.TotalFeesMsat
+			}
+
+			if event.ValueMsat == htlcSum {
+				return &PaymentResp{
+					Preimage: event.PaymentPreimage,
+					Hash:     event.PaymentHash,
+					Status:   Pending,
+				}, nil
+			}
+
+		case lnrpc.Payment_FAILED:
+			return nil, fmt.Errorf("failed payment")
+		}
+	}
+}
+
+// GetPaymentStatus API.
+func (l *LndGrpcLightningAPI) GetPaymentStatus(ctx context.Context, paymentHash string) (*PaymentResp, error) {
+	var err error
+	req := &routerrpc.TrackPaymentRequest{}
+	req.PaymentHash, err = hex.DecodeString(paymentHash)
+	if err != nil {
+		return nil, err
+	}
+	req.NoInflightUpdates = true
+
+	resp, err := l.RouterClient.TrackPaymentV2(ctx, req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout")
+		default:
+			// Do nothing
+		}
+		event, err := resp.Recv()
+
+		if err != nil {
+			return nil, err
+		}
+
+		switch event.Status {
+		case lnrpc.Payment_SUCCEEDED:
+			return &PaymentResp{
+				Preimage: event.PaymentPreimage,
+				Hash:     event.PaymentHash,
+				Status:   Success,
+			}, nil
+		case lnrpc.Payment_FAILED:
+			return &PaymentResp{
+				Preimage: "",
+				Hash:     event.PaymentHash,
+				Status:   Failed,
+			}, nil
+		default:
+			return &PaymentResp{
+				Preimage: "",
+				Hash:     event.PaymentHash,
+				Status:   Pending,
+			}, nil
+		}
+	}
+}
+
+// CreateInvoice API.
+func (l *LndGrpcLightningAPI) CreateInvoice(ctx context.Context, sats int64, preimage string, memo string, expiry time.Duration) (*InvoiceResp, error) {
+	var err error
+	req := &lnrpc.Invoice{}
```

Comment:
I dont like setting everything to zero value and then in next 2 lines change something again